docs_nnx/key_concepts.md

```diff
@@ -0,0 +1,320 @@
+import jax
+import jax.numpy as jnp
+import flax
+from flax import nnx
+from functools import partial
+```
+
+## What is JAX?
+
+JAX is the lower level library that does **all the large-scale data computations**. It provides the singular data container, aka the `jax.Array`, and all the ways we possibly deal with them:
+
+* **Make arithmetic operations upon the arrays**, including: the `jax.numpy` ops, automatic differentiation (`jax.grad`), batching (`jax.vmap`), and more.
+
+* **Run computation on accelerators**, including: interface with various accelerator platforms and layouts; allocating buffers for arrays; compile and execute computation programs across accelerators.
+
+* **Bundle multiple arrays together** using a simple concept called [pytrees](#pytrees).
+
+This implies that any error related with accelerators and numericals are probably a JAX issue, or an issue with Flax built-in layers.
+
+It also means you *can* build a neural network model with JAX alone, especially if you are comfortable with functional programming. JAX docsite have some [simple examples](https://docs.jax.dev/en/latest/notebooks/neural_network_with_tfds_data.html). The article [GPT in 60 Lines of NumPy](https://jaykmody.com/blog/gpt-from-scratch/) also shows how to implement all the key elements of a GPT using JAX.
+
+
+```python
+def jax_linear(x, kernel, bias):
+  return jnp.dot(x, kernel) + bias
+
+params = {'kernel': jax.random.normal(jax.random.key(42), (4, 2)), 
+          'bias': jnp.zeros((2,))}
+x = jax.random.normal(jax.random.key(0), (2, 4))
+y = jax_linear(x, params['kernel'], params['bias'])
+```
+
+## What is Flax?
+
+Flax is a **neural network toolkit**, offering higher level abstractions that are handy for model developers. Such as:
+
+* **Object-oriented `Module` class** to represent layers/models and bookkeep parameters.
+
+* **Modeling utilities** like random number handling, model traversal and surgery, optimizers, advanced parameter bookkeeping, sharding annotations, and more.
+
+* **Some built-in commonly-used** layers, initializers, and model examples.
+
+Take the example below: A Flax layer `Linear`, during initialization, takes one RNG key and automatically initialize all internal parameters as `jax.Array`s. In forward pass, it carries out the exact same computation via JAX APIs.
+
+
+```python
+# Eligible parameters were created inside `linear`, using one RNG key 42
+linear = nnx.Linear(in_features=4, out_features=2, rngs=nnx.Rngs(42))
+
+# Flax created a `Param` wrapper over the actual `jax.Array` parameter to track metadata
+print(type(linear.kernel))        # flax.nnx.Param
+print(type(linear.kernel.value))  # jax.Array
+
+# The computation of the two are the same
+x = jax.random.normal(jax.random.key(0), (2, 4))
+flax_y = linear(x)
+jax_y = jax_linear(x, linear.kernel.value, linear.bias.value)
+assert jnp.array_equal(flax_y, jax_y)
+```
+
+    <class 'flax.nnx.variablelib.Param'>
+    <class 'jaxlib._jax.ArrayImpl'>
+
+
+## Pytrees
```

Comment:
A minor correction in this section is that pytrees don't really know about `Array`s but in the context of Flax it might make sense to focus on them? Else, the generic term would be "leaves".

Reply:
Yeah, I kinda wanted to express in a more novice-friendly way. Rephrasing it to be technically correct.